distributed/client.py

```diff
@@ -3203,7 +3203,10 @@ def who_has(self, futures=None, **kwargs):
             keys = list(map(stringify, {f.key for f in futures}))
         else:
             keys = None
-        return WhoHas(self.sync(self.scheduler.who_has, keys=keys, **kwargs))
+        if self.asynchronous:
+            return self.sync(self.scheduler.who_has, keys=keys, **kwargs)
+        else:
+            return WhoHas(self.sync(self.scheduler.who_has, keys=keys, **kwargs))
```

Comment:
Nitpick: To avoid repeated code we could do something like (note, I've not tested the diff below): 

```suggestion
        result = self.sync(self.scheduler.who_has, keys=keys, **kwargs)
        if self.asynchronous:
            return result
        else:
            return WhoHas(result)
```